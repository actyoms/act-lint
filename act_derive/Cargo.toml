[package]
name = "act_derive"
version = "0.1.0"
edition = "2021"

[lib]
proc-macro = true
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[[test]]
name = "ui"
path = "tests/tests.rs"

[dependencies]
attribute-derive = { version = "0.6.1", features = ["syn-full"] }
proc-macro2 = "1.0.66"
quote = "1.0.32"
serde = { version = "1.0.180", features = ["derive"] }
syn = { version = "2.0.28", features = ["full", "extra-traits"] }
trybuild = { version = "1.0.82", features = ["diff"] }
act_trait = { path = "../act_trait" }
