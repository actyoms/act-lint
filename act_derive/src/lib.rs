extern crate act_macrons;
extern crate act_trait;
extern crate serde;

use proc_macro::TokenStream;

use attribute_derive::Attribute;
use proc_macro2::Ident;
use syn::{Data, DeriveInput, Error, Generics, Visibility};

use deserialize::impl_deserializer;
use parse::parse_input;

mod deserialize;
mod expecting;
mod parse;
mod visitors;

#[derive(Attribute, Debug)]
#[attribute(ident = act)]
struct Attributes {
    #[attribute(conflicts = [visit])]
    visitor: Option<String>,
    #[attribute(conflicts = [visitor])]
    visit: Option<Vec<String>>,
    expecting: Option<String>,
}

struct Input {
    name: Ident,
    attrs: Attributes,
    #[allow(dead_code)]
    vis: Visibility,
    #[allow(dead_code)]
    generics: Generics,
    #[allow(dead_code)]
    data: Data,
    impls: Implementations,
}

#[derive(Default)]
struct Implementations {
    _bool: bool,
    _f64: bool,
    _i64: bool,
    _map: bool,
    _str: bool,
    _u64: bool,
    _number: bool,
}

impl Input {
    fn should_add_i64(&self) -> bool {
        self.impls._i64 || self.impls._number
    }
    fn should_add_u64(&self) -> bool {
        self.impls._u64 || self.impls._i64 || self.impls._number
    }
    fn should_add_f64(&self) -> bool {
        self.impls._f64 || self.impls._number
    }
    fn should_add_str(&self) -> bool {
        self.impls._str
    }
    fn should_add_bool(&self) -> bool {
        self.impls._bool
    }
    fn should_add_map(&self) -> bool {
        self.impls._map
    }
}

impl Input {
    fn from_attributes(attrs: Attributes, drive_input: DeriveInput) -> Input {
        let impls = if let Some(visit) = &attrs.visit {
            Implementations {
                _bool: visit.contains(&"bool".to_string()),
                _f64: visit.contains(&"f64".to_string()),
                _i64: visit.contains(&"i64".to_string()),
                _map: visit.contains(&"map".to_string()),
                _str: visit.contains(&"str".to_string()),
                _u64: visit.contains(&"i64".to_string()),
                _number: visit.contains(&"number".to_string()),
            }
        } else {
            Implementations::default()
        };
        Input {
            name: drive_input.ident,
            attrs,
            vis: drive_input.vis,
            generics: drive_input.generics,
            data: drive_input.data,
            impls,
        }
    }
}

#[proc_macro_derive(Deserialize, attributes(act))]
pub fn deserialize_derive(tokens: TokenStream) -> TokenStream {
    parse_input(tokens)
        .map(impl_deserializer)
        .unwrap_or_else(Error::into_compile_error)
        .into()
}
