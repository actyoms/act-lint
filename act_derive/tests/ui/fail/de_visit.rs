use act_derive::Deserialize;

#[derive(Deserialize)]
#[act(expecting = "something", visit = ["str", "map", "number", "bool"])]
struct MissingVisitImpl;

fn main() {}
