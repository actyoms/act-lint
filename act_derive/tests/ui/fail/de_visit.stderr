error[E0599]: no function or associated item named `from_bool` found for struct `MissingVisitImpl` in the current scope
 --> tests/ui/fail/de_visit.rs:3:10
  |
3 | #[derive(Deserialize)]
  |          ^^^^^^^^^^^ function or associated item not found in `MissingVisitImpl`
4 | #[act(expecting = "something", visit = ["str", "map", "number", "bool"])]
5 | struct MissingVisitImpl;
  | ----------------------- function or associated item `from_bool` not found for this struct
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following trait defines an item `from_bool`, perhaps you need to implement it:
          candidate #1: `FromBool`
  = note: this error originates in the derive macro `Deserialize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no function or associated item named `from_f64` found for struct `MissingVisitImpl` in the current scope
 --> tests/ui/fail/de_visit.rs:3:10
  |
3 | #[derive(Deserialize)]
  |          ^^^^^^^^^^^ function or associated item not found in `MissingVisitImpl`
4 | #[act(expecting = "something", visit = ["str", "map", "number", "bool"])]
5 | struct MissingVisitImpl;
  | ----------------------- function or associated item `from_f64` not found for this struct
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following traits define an item `from_f64`, perhaps you need to implement one of them:
          candidate #1: `FromF64`
          candidate #2: `FromNumber`
  = note: this error originates in the derive macro `Deserialize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no function or associated item named `from_i64` found for struct `MissingVisitImpl` in the current scope
 --> tests/ui/fail/de_visit.rs:3:10
  |
3 | #[derive(Deserialize)]
  |          ^^^^^^^^^^^ function or associated item not found in `MissingVisitImpl`
4 | #[act(expecting = "something", visit = ["str", "map", "number", "bool"])]
5 | struct MissingVisitImpl;
  | ----------------------- function or associated item `from_i64` not found for this struct
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following traits define an item `from_i64`, perhaps you need to implement one of them:
          candidate #1: `FromI64`
          candidate #2: `FromNumber`
  = note: this error originates in the derive macro `Deserialize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no function or associated item named `from_map` found for struct `MissingVisitImpl` in the current scope
 --> tests/ui/fail/de_visit.rs:3:10
  |
3 | #[derive(Deserialize)]
  |          ^^^^^^^^^^^ function or associated item not found in `MissingVisitImpl`
4 | #[act(expecting = "something", visit = ["str", "map", "number", "bool"])]
5 | struct MissingVisitImpl;
  | ----------------------- function or associated item `from_map` not found for this struct
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following trait defines an item `from_map`, perhaps you need to implement it:
          candidate #1: `FromMap`
  = note: this error originates in the derive macro `Deserialize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no function or associated item named `from_str` found for struct `MissingVisitImpl` in the current scope
 --> tests/ui/fail/de_visit.rs:3:10
  |
3 | #[derive(Deserialize)]
  |          ^^^^^^^^^^^ function or associated item not found in `MissingVisitImpl`
4 | #[act(expecting = "something", visit = ["str", "map", "number", "bool"])]
5 | struct MissingVisitImpl;
  | ----------------------- function or associated item `from_str` not found for this struct
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following trait defines an item `from_str`, perhaps you need to implement it:
          candidate #1: `FromStr`
  = note: this error originates in the derive macro `Deserialize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no function or associated item named `from_u64` found for struct `MissingVisitImpl` in the current scope
 --> tests/ui/fail/de_visit.rs:3:10
  |
3 | #[derive(Deserialize)]
  |          ^^^^^^^^^^^ function or associated item not found in `MissingVisitImpl`
4 | #[act(expecting = "something", visit = ["str", "map", "number", "bool"])]
5 | struct MissingVisitImpl;
  | ----------------------- function or associated item `from_u64` not found for this struct
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following traits define an item `from_u64`, perhaps you need to implement one of them:
          candidate #1: `FromNumber`
          candidate #2: `FromU64`
  = note: this error originates in the derive macro `Deserialize` (in Nightly builds, run with -Z macro-backtrace for more info)
